IdentityCore/src/cache/mac/MSIDMacKeychainTokenCache.m

```diff
@@ -443,44 +556,128 @@ - (MSIDCredentialCacheItem *)tokenWithKey:(__unused MSIDCacheKey *)key
 
 // Read one or more credentials from the keychain that match the key (see credentialItem:matchesKey).
 // If not found, return an empty list without setting an error.
-- (NSArray<MSIDCredentialCacheItem *> *)tokensWithKey:(__unused MSIDCacheKey *)key
-                                           serializer:(__unused id<MSIDCacheItemSerializing>)serializer
-                                              context:(__unused id<MSIDRequestContext>)context
-                                                error:(__unused NSError **)error
+- (NSArray<MSIDCredentialCacheItem *> *)tokensWithKey:(MSIDCacheKey *)key
+                                           serializer:(id<MSIDCacheItemSerializing>)serializer
+                                              context:(id<MSIDRequestContext>)context
+                                                error:(NSError **)error
 {
-    NSArray *tokenItems = [self itemsWithKey:key context:context error:error];
-
-    if (!tokenItems)
-    {
-        return nil;
-    }
-
-    NSMutableArray<MSIDCredentialCacheItem *> *tokenList = [NSMutableArray new];
-    for (NSDictionary *dict in tokenItems)
+    NSArray<MSIDCredentialCacheItem *> *tokenList = @[];
+    
+    if ([key isKindOfClass:([MSIDDefaultCredentialCacheKey class])])
     {
-        NSData *jsonData = dict[(id)kSecValueData];
-        if (jsonData)
+        /*
+         Sync in memory cache with persistent cache at the time of look up.
+         */
+        @synchronized (self)
         {
-            MSIDCredentialCacheItem *cacheItem = (MSIDCredentialCacheItem *)[serializer deserializeCredentialCacheItem:jsonData];
-            if (cacheItem != nil)
+            /*
+             For refresh tokens, always merge with persistence to get the most recent refresh token as it is shared across apps from same publisher.
+             For AT/ID tokens, Look up in memory cache first, if no credential found, then merge it with persistent cache and perform additional look up.
+             */
+            MSIDMacCredentialStorageItem *savedItem = [self storageItemWithKey:key serializer:serializer context:context error:error];
+            MSIDMacCredentialStorageItem *currentItem = key.isShared ? self.sharedStorageItem : self.appStorageItem;
+            
+            if (key.isShared)
             {
-                [tokenList addObject:cacheItem];
+                [currentItem mergeStorageItem:savedItem];
```

Comment:
should we do `[savedItem mergeStorageItem:currentItem]`? As the function call prefers the value of the caller :)

Reply:
As discussed. The current logic is there for the case when in-memory refresh token is newer.

Reply:
After talking to Rohit, seems we are not 100% sure whether the RT in keychain or memory is newer, so i am fine with the current implementation :)